drivers/src/pci/capabilities.rs

```diff
@@ -45,21 +47,98 @@ mod header_offsets {
     define_field_span!(CapabilityHeader, next_cap, u8);
 }
 
+mod pmc_offsets {
+    use super::PowerManagementRegisters;
+    use crate::define_field_span;
+
+    define_field_span!(PowerManagementRegisters, pmc, u16);
+    define_field_span!(PowerManagementRegisters, pmcsr, u16);
+}
+
+// Type-specific capability structures.
+#[enum_dispatch]
+enum CapabilityType {
+    PowerManagement,
+}
+
+// Common functionality required by all capabilities.
+#[enum_dispatch(CapabilityType)]
+trait Capability {
+    // Returns the length of the capability, including the common header.
+    fn length(&self) -> usize;
+
+    // Emulates a read from the type-specific registers of this capability.
+    fn emulate_read(&self, op: &mut MmioReadBuilder, cap_offset: usize);
+
+    // Emulates a write to the type-specific registers of this capability.
+    fn emulate_write(&mut self, op: &mut MmioWriteBuilder, cap_offset: usize);
+}
+
+struct PowerManagement {
+    registers: &'static mut PowerManagementRegisters,
+}
+
+impl PowerManagement {
+    fn new(header: &mut CapabilityHeader) -> Self {
+        // Safety: `header` points to a valid and unqiuely-owned capability structure and we are
+        // trusting that the hardware reported the type of the capability correctly.
+        let registers = unsafe {
+            (header as *mut CapabilityHeader as *mut PowerManagementRegisters)
+                .as_mut()
+                .unwrap()
+        };
+        Self { registers }
+    }
+}
+
+impl Capability for PowerManagement {
+    fn length(&self) -> usize {
+        size_of::<PowerManagementRegisters>()
+    }
+
+    fn emulate_read(&self, op: &mut MmioReadBuilder, cap_offset: usize) {
+        use pmc_offsets::*;
+        match cap_offset {
+            pmc::span!() => {
+                op.push_word(self.registers.pmc.readable_bits());
+            }
+            pmcsr::span!() => {
+                op.push_word(self.registers.pmcsr.readable_bits());
+            }
+            _ => {
+                op.push_byte(0);
+            }
+        }
+    }
+
+    fn emulate_write(&mut self, op: &mut MmioWriteBuilder, _cap_offset: usize) {
+        // TODO: Support PME and D-state transitions if necessary.
+        op.pop_byte();
+    }
+}
+
 // Represents a single PCI capability.
 struct PciCapability {
     id: CapabilityId,
     offset: usize,
     next: usize,
+    cap_type: Option<CapabilityType>,
```

Comment:
Can you add a comment that `None` is used as a type for unimplemented capabilities?

Reply:
Done.